Added user facing aliases for AbstractWires and AbstractArray (#9701)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

Reviewer: I'm wondering if we should merge this changleog entry with,
```
* `AbstractArray` has been added to
  `pennylane.typing`, and `AbstractWires` has been added to `pennylane.wires`.
  These will support a new method of having compressed operators for resource estimation
  and decomposition.
  [(#9385)](https://github.com/PennyLaneAI/pennylane/pull/9385)
```
since we're in the same dev cycle. @isaacdevlugt

<h3>Improvements 🛠</h3>

* Type aliases `Int`, `Float`, `Complex`, `Bool`, `Wire` have been introduced that allow for an intuitive 
  sized abstract type notation.  

>   [(#9701)](https://github.com/PennyLaneAI/pennylane/pull/9701)

Reviewer 2: ```suggestion
* Type aliases `Int`, `Float`, `Complex`, `Bool`, `Wire` have been introduced that allow for intuitive 
  abstract type notation.  
  [(#9701)](https://github.com/PennyLaneAI/pennylane/pull/9701)
```

Reviewer 2: Should we describe what these abstract types are meant for, or is it too early at this stage?

Reviewer 2: even some of the code examples in the docstrings for each type here might not be a bad idea at this stage :). E.g., 

```
>>> isinstance(Wires([0, 1]), qp.typing.Wire[2])
True
>>> qp.typing.Wire[2]
AbstractWires(num_wires=2)
```

Author: Added some examples for you!

<h3>Improvements 🛠</h3>

* Type aliases `Int`, `Float`, `Complex`, `Bool`, and `Wire` have been introduced to allow for intuitive 
  abstract type notation.  
  [(#9701)](https://github.com/PennyLaneAI/pennylane/pull/9701)

  ```python
  from pennylane.typing import Int, Float, Complex, Bool, Wire
  Int[2, 3]       # Abstract int array with shape (2, 3)
  Float           # Float scalar
  Complex[...]    # Abstract complex array with any shape
  Bool[-1, 3, 4]  # Abstract bool array with dynamic size for the first axis
  Wire            # Single abstract wire
  Wire[4]         # Four abstract wires
  Wire[-1]        # Wire sequence with dynamic size
  ```
  For example, these abstract types can be used to do type-checking on concrete values:
  
  ```pycon
  >>> isinstance(np.array(False), qp.typing.Bool)
  True
  >>> qp.typing.Bool[4]
  AbstractArray(shape=(4,), dtype=dtype('bool'))
  >>> isinstance(np.array(0+1.2j), qp.typing.Complex)
  True
  >>> qp.typing.Complex[..., 2]
  AbstractArray(shape=(Ellipsis, 2), dtype=dtype('complex128'))
  >>> isinstance(qp.wires.Wires([0, 1]), qp.typing.Wire[2])
  True
  >>> qp.typing.Wire[2]

>   AbstractWires(num_wires=2)

Reviewer: ```suggestion
  AbstractWires(num_wires=2)
  
```

Reviewer: Fixes doc tests

<h3>Improvements 🛠</h3>

> * Type aliases `Int`, `Float`, `Complex`, `Bool`, `Wire` have been introduced that allow for intuitive 

Reviewer 2: ```suggestion
* Type aliases `Int`, `Float`, `Complex`, `Bool`, and `Wire` have been introduced to allow for intuitive 
```

  abstract type notation.  
  [(#9701)](https://github.com/PennyLaneAI/pennylane/pull/9701)

  ```python
  from pennylane.typing import Int, Float, Complex, Bool, Wire
  Int[2, 3]       # Abstract int array with shape (2, 3)
  Float           # Float scalar
  Complex[...]    # Abstract complex array with any shape
  Bool[-1, 3, 4]  # Abstract bool array with dynamic size for the first axis
  Wire            # Single abstract wire
  Wire[4]         # Four abstract wires
  Wire[-1]        # Wire sequence with dynamic size
  ```
 
  ```pycon
  >>> isinstance(np.array(False), qp.typing.Bool)
  True
  >>> qp.typing.Bool[4]
  AbstractArray(shape=(4,), dtype=dtype('bool'))
  >>> isinstance(np.array(0+1.2j), qp.typing.Complex)
  True
  >>> qp.typing.Complex[..., 2]
  AbstractArray(shape=(Ellipsis, 2), dtype=dtype('complex128'))
  >>> isinstance(qp.wires.Wires([0, 1]), qp.typing.Wire[2])
  True
  >>> qp.typing.Wire[2]
  AbstractWires(num_wires=2)

  ```

>   etc.

Reviewer 2: ````suggestion
  For example, these abstract types can be used to do type-checking on concrete values:
  
  ```pycon
  >>> isinstance(np.array(False), qp.typing.Bool)
  True
  >>> qp.typing.Bool[4]
  AbstractArray(shape=(4,), dtype=dtype('bool'))
  >>> isinstance(np.array(0+1.2j), qp.typing.Complex)
  True
  >>> qp.typing.Complex[..., 2]
  AbstractArray(shape=(Ellipsis, 2), dtype=dtype('complex128'))
  >>> isinstance(qp.wires.Wires([0, 1]), qp.typing.Wire[2])
  True
  >>> qp.typing.Wire[2]
  AbstractWires(num_wires=2)
  ```
````

<h3>Improvements 🛠</h3>

* Type aliases `Int`, `Float`, `Complex`, `Bool`, `Wire` have been introduced that allow for intuitive 
  abstract type notation.  
  [(#9701)](https://github.com/PennyLaneAI/pennylane/pull/9701)

  ```python
  from pennylane.typing import Int, Float, Complex, Bool, Wire
  Int[2, 3]       # Abstract int array with shape (2, 3)
  Float           # Float scalar
  Complex[...]    # Abstract complex array with any shape
  Bool[-1, 3, 4]  # Abstract bool array with dynamic size for the first axis
  Wire            # Single abstract wire
  Wire[4]         # Four abstract wires
  Wire[-1]        # Wire sequence with dynamic size
  ```
 
  ```pycon
  >>> isinstance(np.array(False), qp.typing.Bool)
  True

  ```

  ```pycon
  >>> qp.typing.Bool[4]
  AbstractArray(shape=(4,), dtype=dtype('bool'))
  
  ```

  ```pycon
  >>> isinstance(np.array(0+1.2j), qp.typing.Complex)
  True
  
  ```
  
  ```pycon
  >>> qp.typing.Complex[..., 2]
  AbstractArray(shape=(Ellipsis, 2), dtype=dtype('complex128'))
  
  ```

  ```pycon
  >>> isinstance(Wires([0, 1]), qp.typing.Wire[2])
  True
  
  ```
  
  ```pycon
  >>> qp.typing.Wire[2]
  AbstractWires(num_wires=2)

  ```

>   etc.

Reviewer 3: Why not do all these in a single `pycon` code block rather than separate ones?

Author: Sure

Author: https://github.com/PennyLaneAI/pennylane/pull/9701/commits/e6568aca6ed83effc94167e8cbd9d7ba0cec8cd6
